cdap-api/src/main/java/co/cask/cdap/api/lineage/operation/Join.java

```diff
@@ -0,0 +1,128 @@
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.api.lineage.operation;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * Represents join operation.
+ */
+public class Join extends Operation {
+  private final Constraint constraint;
+  private final Type joinType;
+  private final List<String> outputs;
+
+  /**
+   * Create an instance of join operation.
+   * @param name the name of the join operation
+   * @param description the description of the join operation
+   * @param constraint the constraint for the join operation
+   * @param joinType the type of the join being performed
+   * @param outputs the array of outputs created by the join operation. This array simply contains
+   *                the outputs from both the entities on which join operation is being performed,
+   *                since join operation itself does not change any output. So it is assumed that
+   *                all outputs from the joined entities should have unique name.
+   */
+  public Join(String name, String description, Constraint constraint, Type joinType, String... outputs) {
+    this(name, description, constraint, joinType, Arrays.asList(outputs));
+  }
+
+  /**
+   * Create an instance of join operation.
+   * @param name the name of the join operation
+   * @param description the description of the join operation
+   * @param constraint the constraint for the join operation
+   * @param joinType the type of the join being performed
+   * @param outputs the list of outputs created by the join operation. This array simply contains
+   *                the outputs from both the entities on which join operation is being performed,
+   *                since join operation itself does not change any output. So it is assumed that
+   *                all outputs from the joined entities should have unique name.
+   */
+  public Join(String name, String description, Constraint constraint, Type joinType, List<String> outputs) {
+    super(name, co.cask.cdap.api.lineage.operation.Type.JOIN, description);
+    this.joinType = joinType;
+    this.constraint = constraint;
+    this.outputs = Collections.unmodifiableList(new ArrayList<>(outputs));
+  }
+
+  /**
+   * @return join constraint
+   */
+  public Constraint getConstraint() {
+    return constraint;
+  }
+
+  /**
+   * @return the type of join being performed
+   */
+  public Type getJoinType() {
+    return joinType;
+  }
+
+  /**
+   * @return list of outputs created by the join operation
+   */
+  public List<String> getOutputs() {
+    return outputs;
+  }
+
+  /**
+   * Enum to hold different join types
+   */
+  public enum Type {
+    INNER,
+    OUTER,
+    LEFT,
+    RIGHT
+  }
+
+  /**
+   * Class to represent the join constraint
+   */
+  public static class Constraint {
+    List<Input> left;
```

Comment:
I think this is too generic. Using this API you could use input fields from various different origins, whereas a join really only has two inputs, the left and the right side of the join. See my comment above.

Reply:
added javadoc

Reply:
actually I was wrong, it must be possible to have fields from different origins.